core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -138,9 +138,11 @@ class ControllerChannelManager(controllerContext: ControllerContext, config: Kaf
 
   private def removeExistingBroker(brokerState: ControllerBrokerStateInfo) {
     try {
+      // shutdown the RequestSendThread before closing the NetworkClient to
+      // avoid the concurrent use of the non-threadsafe classes as described in KAFKA-4959
+      brokerState.requestSendThread.shutdown()
```

Comment:
Concurrent access to messageQueue between the ZkEventThread and RequestSendThread is safe because it's a `java.util.concurrent.LinkedBlockingQueue` which is thread-safe.

I can clarify all of this in the comments.